### DISCUSSION TITLE
Support for multiple UI libraries using composable components

### DISCUSSION THREAD

**Discussion by jgbernalp** [perses/perses#1069] [Ideas]
**BACKGROUND**
Red Hat initiated an evaluation of Perses to provide customizable Dashboards inside the OpenShift web console. As we want to use [Patternfly](https://www.patternfly.org/v4/), our own UI component library. The approach we decided to take was to embed the panels, using the data connectors to the Perses backend and other controls based on our component library. During our evaluation  we found that the panels and other components we could reuse are coupled with Material UI, a re-skinning is not possible as the UX patterns differ and our bundle size will increase.

**GOAL**
Allow users to bring their own UI components when embedding Perses components, charts and front-end logic.

**PROPOSAL**
Decouple embeddable components such as panels from a specific component library by providing them with UI.

For example `/ui/dashboards/src/components/Panel/Panel.tsx`:
```typescript
import { Card, CardContent } from '@mui/material';

export function Panel(props: PanelProps) {
  return <Card>
     <PanelHeader />
     <CardContent>
      <PanelContent />
     </CardContent>
  </Card>
}
```
will change to:
```typescript
export function Panel(props: PanelProps) {
  // A "panel" context key is passed to return the Card component instance specific for panels
  const { Card, CardContent } = useCustomComponents("panel"); 

  return <Card>
     <PanelHeader />
     <CardContent>
      <PanelContent />
     </CardContent>
  </Card>
}
```
```typescript
interface CustomComponents {
   panel: Record<"Card" | "CardContent", React.Component>
}

// or a more specific interface if props from the parent component are needed
interface CustomComponents {
   panel: {
    Card: React.Component<CardProps>,
    CardContent: React.Component<CardContentProps>,
}

export const CustomComponentsContext = createContext<CustomComponents | undefined>(undefined);

export function CustomComponentsProvider(props: CustomComponentsProviderProps) {
  const { components } = props;
  return <CustomComponentsContext.Provider value={components}>{children}</CustomComponentsContext.Provider>;
}
```

**RISKS**
- Theming: will provided components respect the theme created for the specific UI library?. The provided components should be enclosed with the theme provider for the UI library.
- Props: what happens when the UI library components require properties from the parent component?. An interface can be defined so the provided component implements it when needed, but initially this is aimed for core components to ease usage when embedding Perses and not for a full UI library replacement.
- File structure: where should the UI library component instances be placed for the default MUI?. In the example above, we could co-locate the Card specific to the panel with the Panel.tsx so is easy to maintain and rely on tree shaking so is not included when embedding.

**Comment by sjcobb** [comment]
This is a really creative solution overall that we hadn’t thought of, I’m curious how far have you gone when testing this as part of the spike? Would it cause any issues with Webpack / how our libraries are bundled, particularly MUI code-splitting?

**Reply by jgbernalp** [reply]
We initially replaced all the MUI components with ours, but we have the feeling that they will start to diverge if we choose that path. This means, we cannot reuse at least the core components like the Panel. I trust Webpack does its job to split this component in a chunk away from MUI, as it does not depend on it anymore, and put the MUI provider along with the MUI related code.

**Comment by juliepagano** [comment]
Do have a sense of which parts of the Perses packages you would need this approach? For example, do you want to be able to override content from plugins like the settings that show up when you edit the panel and the inner contents of panels? Getting a sense of the scale of what you need would be really helpful for evaluating the proposal.

**Reply by jgbernalp** [reply]
We are working initially on a read-only view, so panels is the current scope.

**Reply by juliepagano** [reply]
Just to clarify, is it just the panel wrapper (e.g. header, footer, border) you want to modify OR do you also potentially want to modify the internals of the panel for some plugins (e.g. changing how a time series looks).

**Reply by jgbernalp** [reply]
The goal is to evaluate if we could use Perses components, panels and charts, without MUI. The reasons behind it is that we want to use our UI components and keep a small bundle size.

**Reply by sjcobb** [reply]
That's helpful, at the beginning of the spike, I thought the goal was the ability to style your version of a Perses dashboard to look similar to Patternfly and that bundle size was a less important concern. It sounds like now, even if we had a much better way to do that on top of the MUI theme and ChartsThemeProvider, that wouldn't be sufficient?

Overall, using our components, core panels, and higher level composite components is a big value proposition of Perses. We spent a lot of time on certain functionality like our custom tooltip, we have plans to build a better legend and time range picker, the core panels have a lot of thoughtful customization (thanks to MUI), and we are constantly iterating on panel options. One concern I have is what you build now is going to be out of date with our dashboard and [panel spec](https://github.com/perses/perses/tree/v0.27.0/schemas/panels) very quickly.

**Comment by jgbernalp** [comment]
While testing this approach I realized the plugins are also dependent on MUI for editing as @juliepagano mentioned above. So in order to load data from prometheus MUI is also included. This also aligns with https://github.com/perses/perses/discussions/956.

**Comment by jgbernalp** [comment]
Coming back to this topic. The initial proposal about using a provider for UI components seems to be cumbersome and it might cause a big maintenance effort. We want to keep in line with the dashboard and panel spec, not building something that deviates us from it. Leveraging browser cache when having Material UI loaded with Patternfly will reduce the impact on the client side. I'll close this one and check on alternatives, such as loading only parts of the plugin for read-only dashboards excluding the editing components.
